lib/rule-translation/src/RuleTranslation.hs

```diff
@@ -0,0 +1,449 @@
+                                                                   varDoc = text newVar
+
+makeDestructorDefinition :: (Show l) => Term l -> String
+makeDestructorDefinition t = 
+  "forall " ++ intercalate ", " (map (++":bitstring") atoms) ++ ";#" ++ printTerm2 t
+  where
+    atoms = map showAtom2 . S.toList . S.fromList $ getAtoms t
+
+makeVariable :: (Show l) => Term l -> M.Map String String -> (String, M.Map String String)
+makeVariable t varMap = case M.lookup (printTerm S.empty False t) varMap of
+    Just v  -> (v, varMap)
+    Nothing -> let newVar = "helperVar" ++ (show $ M.size varMap)
+                   newMap = M.insert (printTerm S.empty False t) newVar varMap
+                 in
+               (newVar, newMap)
+      
+
+makeDestructorName :: (Show l) => M.Map (String, String) String -> Term l -> String -> (String, M.Map (String, String) String)
+makeDestructorName dMap t a = case M.lookup ((makeDestructorDefinition t),a) dMap of
+    Just d  -> (d, dMap)
+    Nothing -> let newDestructor = "g_" ++ (showAtom2 a) ++ "_" ++ (show $ M.size dMap)
+                   newMap = M.insert ((makeDestructorDefinition t),a) newDestructor dMap
+                 in
+               (newDestructor, newMap)
+
+makeDestructorExpressions :: (Document d, Show l, Ord l) => S.Set String -> M.Map String String -> M.Map (String, String) String -> Term l -> (d, S.Set String, M.Map (String, String) String)
+makeDestructorExpressions vars helperVars destructors t = 
+    ((vcat doclist), S.fromList atoms, newDestructors)
+    where
+      (doclist, newDestructors) = foldl (\(docs,destrs) a -> let (doc, destrs') = makeDestructorExpression vars helperVars destrs t a in (docs ++ [doc], destrs')) ([], destructors) atoms
+      atoms = nub $ getAtoms t
+
+getAtoms :: (Show l) => Term l -> [String]
+getAtoms t = case viewTerm t of
+    Lit l     -> [show l]
+    FApp _ ts -> foldl (\acc t -> acc ++ getAtoms t) [] ts
+
+makeDestructorExpression :: (Document d, Show l) => S.Set String -> M.Map String String -> M.Map (String, String) String -> Term l -> String -> (d, M.Map (String, String) String)
+makeDestructorExpression vars helperVars destructors t a = if (S.member a vars) || (head a == '\'')
+                                      then (oldVarDoc, newDestructors)
+                                      else (newVarDoc, newDestructors)
+                                      where
+                                        (var, _) = makeVariable t helperVars
+                                        (destr, newDestructors) = makeDestructorName destructors t a
+                                        oldVarDoc = text "let (=" <> (text $ showAtom a) <>
+                                                text ") =" <-> text destr <> 
+                                                text "(" <> text var <> text ") in"
+                                        newVarDoc = text "let" <-> (text $ showAtom a) <->
```

Comment:
rephrase so the difference between oldVarDoc and NewVarDoc become more clear (e.g., put the conditional inside the definition of the doc [of course abbreviated]
)